File: apps/mobile-android/app/src/main/kotlin/br/com/tatame/core/theme/ThemeStore.kt
package br.com.tatame.core.theme

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import br.com.tatame.core.designsystem.palette.BrandInput
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

/**
 * Device-local theme persistence (spec 011, CFG.14/15):
 * - last-seen academy brand triplet, so a cold start paints branded before
 *   the session refresh lands (null = default Tatame brand);
 * - the explicit "Tema escuro" preference (per-user, client-side — design
 *   system ticket 06; default light, system-follow is recorded debt).
 */
interface ThemeStore {
    /** Emits the persisted dark preference (false when never set). */
    val darkTheme: Flow<Boolean>

    suspend fun readBrand(): BrandInput?

    /** null clears the cache (logout / academy without branding). */
    suspend fun writeBrand(brand: BrandInput?)

    suspend fun writeDarkTheme(enabled: Boolean)
}

/**
 * [ThemeStore] over Preferences DataStore (TokenStore pattern: only DataStore
 * *core* types referenced, so the class is JVM-unit-testable). The triplet is
 * stored as three keys written atomically in one `edit`; a partial triplet
 * (impossible via [writeBrand], conceivable via corruption) reads as null —
 * fail-closed to the default brand.
 */
class DataStoreThemeStore(
    private val dataStore: DataStore<Preferences>,
) : ThemeStore {

    override val darkTheme: Flow<Boolean> =
        dataStore.data.map { it[KEY_DARK_THEME] ?: false }

    override suspend fun readBrand(): BrandInput? {
        val prefs = dataStore.data.first()
        val deep = prefs[KEY_BRAND_DEEP] ?: return null
        val vibrant = prefs[KEY_BRAND_VIBRANT] ?: return null
        val accent = prefs[KEY_BRAND_ACCENT] ?: return null
        return BrandInput(deep = deep, vibrant = vibrant, accent = accent)
    }

    override suspend fun writeBrand(brand: BrandInput?) {
        dataStore.edit { prefs ->
            if (brand == null) {
                prefs.remove(KEY_BRAND_DEEP)
                prefs.remove(KEY_BRAND_VIBRANT)
                prefs.remove(KEY_BRAND_ACCENT)
            } else {
                prefs[KEY_BRAND_DEEP] = brand.deep
                prefs[KEY_BRAND_VIBRANT] = brand.vibrant
                prefs[KEY_BRAND_ACCENT] = brand.accent
            }
        }
    }

    override suspend fun writeDarkTheme(enabled: Boolean) {
        dataStore.edit { it[KEY_DARK_THEME] = enabled }
    }

    private companion object {
        val KEY_BRAND_DEEP = stringPreferencesKey("brand_deep")
        val KEY_BRAND_VIBRANT = stringPreferencesKey("brand_vibrant")
        val KEY_BRAND_ACCENT = stringPreferencesKey("brand_accent")
        val KEY_DARK_THEME = booleanPreferencesKey("dark_theme")
    }
}

File: apps/mobile-rn/src/theme/theme-store.ts
/**
 * Theme preference store (CFG.12/13, spec 011): module singleton (mirrors
 * session-store) holding the two client-side theming facts —
 *
 * - `brand`: the academy 3-color white-label brand. Sourced from the
 *   session (`academy.theme`) while authed; AsyncStorage-cached so a cold
 *   start paints branded before the session resolves; `null` = default
 *   Tatame brand (logged out, unbranded academy, plataforma).
 * - `mode`: the per-user, per-device dark preference (design-system ticket
 *   06 — client-side only, no server involvement), persisted in
 *   AsyncStorage.
 */

import { useSyncExternalStore } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import type { BrandInput, Mode } from '@tatame/design-system/native';

export const BRAND_CACHE_KEY = 'tatame.lastBrand';
export const THEME_MODE_KEY = 'tatame.themeMode';

export interface ThemeState {
  mode: Mode;
  brand: BrandInput | null;
}

let state: ThemeState = { mode: 'light', brand: null };
/** Once a session verdict lands, the stale cold-start cache must not win. */
let brandResolvedFromSession = false;
const listeners = new Set<() => void>();

function setState(next: ThemeState): void {
  state = next;
  for (const listener of [...listeners]) listener();
}

function subscribe(listener: () => void): () => void {
  listeners.add(listener);
  return () => listeners.delete(listener);
}

function getSnapshot(): ThemeState {
  return state;
}

export function useThemePreferences(): ThemeState {
  return useSyncExternalStore(subscribe, getSnapshot, getSnapshot);
}

/* ------------------------------------------------------------------ */
/* Persistence                                                         */
/* ------------------------------------------------------------------ */

const HEX = /^#[0-9A-Fa-f]{6}$/;

/** Cached triplet must be a full valid brand — anything else is default. */
function parseCachedBrand(raw: string | null): BrandInput | null {
  if (!raw) return null;
  try {
    const value = JSON.parse(raw) as Partial<BrandInput> | null;
    if (
      value &&
      HEX.test(value.deep ?? '') &&
      HEX.test(value.vibrant ?? '') &&
      HEX.test(value.accent ?? '')
    ) {
      return {
        deep: value.deep!,
        vibrant: value.vibrant!,
        accent: value.accent!,
      };
    }
  } catch {
    // Corrupt cache === no cache.
  }
  return null;
}

function sameBrand(a: BrandInput | null, b: BrandInput | null): boolean {
  if (a === b) return true;
  if (!a || !b) return false;
  return a.deep === b.deep && a.vibrant === b.vibrant && a.accent === b.accent;
}

/**
 * Cold-start hydration (before `boot()`): last-seen brand + persisted mode
 * from AsyncStorage, so the splash/first frame already wear the academy's
 * colors and the chosen mode. Unreadable storage === defaults.
 */
export async function hydrateTheme(): Promise<void> {
  let mode: Mode = state.mode;
  let brand: BrandInput | null = state.brand;
  try {
    const [storedMode, storedBrand] = await Promise.all([
      AsyncStorage.getItem(THEME_MODE_KEY),
      AsyncStorage.getItem(BRAND_CACHE_KEY),
    ]);
    if (storedMode === 'dark' || storedMode === 'light') mode = storedMode;
    if (!brandResolvedFromSession) brand = parseCachedBrand(storedBrand);
  } catch {
    // Defaults stand.
  }
  if (mode !== state.mode || !sameBrand(brand, state.brand)) {
    setState({ mode, brand });
  }
}

/**
 * Session verdict → active brand + cache. Authed with a branded academy
 * caches the triplet; authed unbranded / logged out clears back to the
 * default Tatame brand (logout must never leave a stale tenant brand).
 */
export async function applySessionBrand(
  brand: BrandInput | null,
): Promise<void> {
  brandResolvedFromSession = true;
  const next = brand
    ? { deep: brand.deep, vibrant: brand.vibrant, accent: brand.accent }
    : null;
  if (!sameBrand(next, state.brand)) setState({ ...state, brand: next });
  try {
    if (next) await AsyncStorage.setItem(BRAND_CACHE_KEY, JSON.stringify(next));
    else await AsyncStorage.removeItem(BRAND_CACHE_KEY);
  } catch {
    // Cache write failure only costs the next cold start's first frame.
  }
}

/** Aluno perfil "Tema escuro" switch — flips + persists the mode. */
export async function setThemeMode(mode: Mode): Promise<void> {
  if (mode !== state.mode) setState({ ...state, mode });
  try {
    await AsyncStorage.setItem(THEME_MODE_KEY, mode);
  } catch {
    // Persistence failure keeps the in-session flip.
  }
}

/* ------------------------------------------------------------------ */
/* Test seam (rn-07 pattern)                                           */
/* ------------------------------------------------------------------ */

export const themeTestApi = {
  reset(): void {
    state = { mode: 'light', brand: null };
    brandResolvedFromSession = false;
    listeners.clear();
  },
};
